test/unit/test-action-allowlist.js

```diff
@@ -0,0 +1,446 @@
+            '{"tagOrTarget":"amp-element","method":"defaultAction"}].'
+        );
+      });
+
+      it('should allow adding actions to the whitelist', () => {
+        const i = getActionInvocation(target, 'print', 'AMP');
+        action.addToWhitelist('AMP', 'print');
+        action.invoke_(i);
+        expect(spy).to.be.calledWithExactly(i);
+      });
+    });
+
+    it('should not allow any action with empty whitelist', () => {
+      action = new ActionService(env.ampdoc, env.win.document);
+      action.setWhitelist([]);
+      const i = getActionInvocation(target, 'print', 'AMP');
+      env.sandbox.stub(action, 'error_');
+      expect(action.invoke_(i)).to.be.null;
+      expect(action.error_).to.be.calledWith(
+        '"AMP.print" is not whitelisted [].'
+      );
+    });
+
+    it('should ignore unparseable whitelist entries', () => {
+      action = new ActionService(env.ampdoc, env.win.document);
+      action.setWhitelist([
+        {tagOrTarget: 'AMP', method: 'pushState'},
+        {invalidEntry: 'invalid'},
+        {},
+        {tagOrTarget: 'AMP', method: 'setState'},
+        {tagOrTarget: '*', method: 'show'},
+        {tagOrTarget: '*'},
+        {method: 'show'},
+      ]);
+      expect(action.whitelist_).to.deep.equal([
+        {tagOrTarget: 'AMP', method: 'pushState'},
+        {tagOrTarget: 'AMP', method: 'setState'},
+        {tagOrTarget: '*', method: 'show'},
+      ]);
+      const i = getActionInvocation(target, 'setState', 'AMP');
+      action.invoke_(i);
+      expect(spy).to.be.calledWithExactly(i);
+    });
+
+    describe('email documents', () => {
+      beforeEach(() => {
+        env.win.document.documentElement.setAttribute('amp4email', '');
+        action = new ActionService(env.ampdoc, env.win.document);
+      });
+
+      it('should supply default actions allowlist', () => {
+        const i = getActionInvocation(target, 'toggleClass', 'AMP');
+        action.invoke_(i);
+        expect(spy).to.be.calledWithExactly(i);
+      });
+
+      it('should not allow non-default actions', () => {
+        const i = getActionInvocation(target, 'print', 'AMP');
+        env.sandbox.stub(action, 'error_');
+        expect(action.invoke_(i)).to.be.null;
+        expect(action.error_).to.be.calledWith(
+          '"AMP.print" is not whitelisted ' +
+            '[{"tagOrTarget":"AMP","method":"setState"},' +
+            '{"tagOrTarget":"*","method":"focus"},' +
+            '{"tagOrTarget":"*","method":"hide"},' +
+            '{"tagOrTarget":"*","method":"show"},' +
+            '{"tagOrTarget":"*","method":"toggleClass"},' +
+            '{"tagOrTarget":"*","method":"toggleVisibility"}].'
+        );
+      });
+
+      describe('default component actions', () => {
+        beforeEach(() => {
+          env.sandbox.stub(Services, 'actionServiceForDoc').returns(action);
+        });
+
+        afterEach(() => {
+          Services.actionServiceForDoc.restore();
```

Comment:
Ditto, the point of a sandbox is you don't need to worry about this for stubs/mocks/fakes.